src/libopensc/card-starcos.c

```diff
@@ -280,6 +281,54 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
 	return encoding;
 }
 
+/**
+ * Clears logged in state for all pins
+ */
+static void starcos_pin_clear_logged_in(sc_card_t *card) {
+	starcos_ex_data * ex_data = (starcos_ex_data*)card->drv_data;
+	memset(&ex_data->pin_logged_in, 0, sizeof(ex_data->pin_logged_in));
+}
+
+/**
+ * Set logged in state for the given pin reference.
+ * Returns -1 if the pin state cannot be set
+ * Note: pin reference must be greater than 0
+ */
+static int starcos_pin_set_logged_in(sc_card_t *card, unsigned char pin_reference) {
```

Comment:
I'm not sure how useful it is to cache the PIN verification status at the driver level. There may be other processes interfering with the card, especially in a long running process such as in the Browser.

I think it should be OK to just leave out the pin status query. The upper software layers (PKCS#11, minidriver) should handle this correctly...

Reply:
Yes, the minidriver can handle this, but some PKCS#11 clients, especially Firefox, cannot "Log in" with StarCOS 3.5: it happens to verify the PIN successfully, but forgets this status immediately and tries to query it. 
This means client side SSL authentication is not possible without this patch.

Reply:
And you are rigth, sorry for the bad indenting. Seems my development environments unter Windows and under Linux do not match, although both are VS Code. I am goint to fix this as well!